gboard_autopsy.py

```diff
@@ -0,0 +1,123 @@
+        return "Extracts Android GBoard app data."
+
+    def getModuleVersionNumber(self):
+        return "1.0"
+
+    def isDataSourceIngestModuleFactory(self):
+        return True
+
+    def createDataSourceIngestModule(self, ingestOptions):
+        return GboardDataSourceIngestModule()
+
+class GboardDataSourceIngestModule(DataSourceIngestModule):
+
+    _logger = Logger.getLogger(GboardDataSourceIngestModuleFactory.moduleName)
+
+    GBOARD_CLIPBOARD_ARTIFACT = 'GBOARD_CLIPBOARD_OBJECT'
+    GBOARD_PACKAGE_NAME = 'com.google.android.inputmethod.latin'
+
+    def log(self, level, msg):
+        self._logger.logp(level, self.__class__.__name__, inspect.stack()[1][3], msg)
+
+    def __init__(self):
+        self.context = None
+
+    def startUp(self, context):
+        """Where module setup and configuration is done
+
+        Args:
+        context (IngestJobContext): Context of the ingest module
+        """
+
+        self.log(Level.INFO, "GBoard Data Source Ingest Module is starting up...")
+
+        # setup the current context
+        self.context = context
+
+        # check if executable exists
+        module_path = os.path.dirname(os.path.abspath(__file__))
+        exe_ext = '.exe' if os.name == 'nt' else ''
+        path_to_exe = os.path.join(module_path, 'gboard-forensics' + exe_ext)
+
+        if not os.path.exists(path_to_exe):
+            raise IngestModuleException('Executable file not found!')
+
+        try:
+            current_case = Case.getCurrentCaseThrows()
+
+            # create artifacts
+            self.clipboard_type = self.createCustomArtifactType(current_case, self.GBOARD_CLIPBOARD_ARTIFACT, 'Gboard Clipboard')
+        except NoCurrentCaseException as ex:
+            self.log(Level.WARNING, "No case currently open. " + ex)
+
+    def process(self, data_source, progress_bar):
+        """Where the analysis is done
+
+        Args:
+        dataSource (Content): autopsy case data source
+        progressBar (DataSourceIngestModuleProgress): progress bar related to this analysis
+        """
+
+        progress_bar.switchToIndeterminate()
+        self.log(Level.INFO, "GBoard Data Source Ingest Module is analysing...")
+
+        try:
+            current_case = Case.getCurrentCaseThrows()
+            services = current_case.getServices()
+
+            blackboard = services.getBlackboard()
+            file_manager = services.getFileManager()
+
+            files = file_manager.findFiles('%', self.GBOARD_PACKAGE_NAME + '/files/clipboard_image')
+            for file in files:
+                self.log(Level.INFO, file.toString())
+                artifact = file.newArtifact(self.clipboard_type.getTypeID())
+
+                try:
+                    # index the artifact for keyword search
+                    blackboard.indexArtifact(artifact)
+                except Blackboard.BlackboardException as ex:
+                    self.log(Level.SEVERE, "Error indexing artifact " + artifact.getDisplayName())
```

Comment:
I would say it wont work that way because `with` uses `__enter__()` and `__exit__()` function from python objects and this objects came from Java, which is an ArrayList<>. Also, this are not resources, those file objects are only metadata information from the file manager db.

Reply:
I'm not too familiar with how this works internally within python. But you're right, if it comes from Java objects then it's fine as is.